### PR TITLE
feat: add initial implementation of `fft/base/fftpack`

### PR DESCRIPTION
**Pull request by gunjjoshi**
Progresses #3131

## Description

> What is the purpose of this pull request?

This pull request:

-   adds `fft/base/fftpack`.
-   adds the following packages:

    - [ ] cfftb
    - [ ] cfftf
    - [ ] cffti
    - [ ] cosqb
    - [ ] cosqf
    - [ ] cosqi
    - [ ] cost
    - [ ] costi
    - [x] decompose
    - [ ] rfftb
    - [x] rfftf
    - [x] rffti
    - [ ] sinqb
    - [ ] sinqf
    - [ ] sinqi
    - [ ] sint
    - [ ] sinti

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   progresses #3131.

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

File: lib/node_modules/@stdlib/fft/base/fftpack/lib/passfb.js [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+	idot = ido / 2;
+	ipp2 = ip + 2;
+	ipph = ( ip + 1 ) / 2;
+	idp = ip * ido;
+	if ( ido >= l1 ) {
+		for ( j = 2; j <= ipph; ++j ) {
+			jc = ipp2 - j;
+			for ( k = 1; k <= l1; ++k ) {
+				for ( i = 1; i <= ido; ++i ) {
+					ch[ chRef( i, k, j, l1, ido ) - ch_offset ] = cc[ ccRef( i, j, k, ip, ido ) - cc_offset ] + cc[ ccRef( i, jc, k, ip, ido ) - cc_offset ]; // eslint-disable-line max-len
```
Comment (kgryte) [inline lib/node_modules/@stdlib/fft/base/fftpack/lib/passfb.js]: Just disable this lint rule at the top-level, so you are not repeatedly having to disable it.

File: lib/node_modules/@stdlib/fft/base/fftpack/lib/passfb.js [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+var ccRef = require( './cc_ref.js' );
+
+
+// MAIN //
+
+/**
+* Performs a pass of the FFT algorithm.
+*
+* @private
+* @param {Float64Array} nac - Number of FFT passes.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/fft/base/fftpack/lib/passfb.js]: No periods at the end of descriptions. Lowercase the first word of descriptions. Applies here and elsewhere in this PR.

File: lib/node_modules/@stdlib/fft/base/fftpack/lib/passb2.js [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+	var i;
+	var k;
+
+	// Parameter adjustments...
+	chOffset = 1 + ( ido * ( 1 + l1 ) );
+	ccOffset = 1 + ( ido * 3 );
+
+	// Function body:
+	if ( ido <= 2 ) {
+		for ( k = 1; k <= l1; ++k ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/fft/base/fftpack/lib/passb2.js]: Is `++k` intentional? If not, we generally prefer `k++`.

File: lib/node_modules/@stdlib/fft/base/fftpack/lib/passb3.js [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+			ti2 = cc[ ccRef( 2, 2, k, 3, ido ) - ccOffset ] + cc[ ccRef( 2, 3, k, 3, ido ) - ccOffset ];
+			ci2 = cc[ ccRef( 2, 1, k, 3, ido ) - ccOffset ] + ( taur * ti2 );
+			ch[ chRef( 2, k, 1, l1, ido ) - chOffset ] = cc[ ccRef( 2, 1, k, 3, ido ) - ccOffset ] + ti2;
+			cr3 = taui * ( cc[ ccRef( 1, 2, k, 3, ido ) - ccOffset ] - cc[ ccRef( 1, 3, k, 3, ido ) - ccOffset ] );
+			ci3 = taui * ( cc[ ccRef( 2, 2, k, 3, ido ) - ccOffset ] - cc[ ccRef( 2, 3, k, 3, ido ) - ccOffset ] );
+			ch[ chRef( 1, k, 2, l1, ido ) - chOffset ] = cr2 - ci3;
+			ch[ chRef( 1, k, 3, l1, ido ) - chOffset ] = cr2 + ci3;
+			ch[ chRef( 2, k, 2, l1, ido ) - chOffset ] = ci2 + ci3;
+			ch[ chRef( 2, k, 3, l1, ido ) - chOffset ] = ci2 - cr3;
+		}
```
Comment (kgryte) [inline lib/node_modules/@stdlib/fft/base/fftpack/lib/passb3.js:131]: You can use an early `return` here so you can remove the `else` and reduce indentation.

File: lib/node_modules/@stdlib/fft/base/fftpack/lib/passb4.js [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+			ch[ chRef( 1, k, 3, l1, ido ) - chOffset ] = tr2 - tr3;
+			ch[ chRef( 2, k, 1, l1, ido ) - chOffset ] = ti2 + ti3;
+			ch[ chRef( 2, k, 3, l1, ido ) - chOffset ] = ti2 - ti3;
+			ch[ chRef( 1, k, 2, l1, ido ) - chOffset ] = tr1 + tr4;
+			ch[ chRef( 1, k, 4, l1, ido ) - chOffset ] = tr1 - tr4;
+			ch[ chRef( 2, k, 2, l1, ido ) - chOffset ] = ti1 + ti4;
+			ch[ chRef( 2, k, 4, l1, ido ) - chOffset ] = ti1 - ti4;
+		}
+	} else {
+		for ( k = 1; k <= l1; ++k ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/fft/base/fftpack/lib/passb4.js]: Same comments regarding early `return` and `++k`.

File: lib/node_modules/@stdlib/fft/base/fftpack/lib/passf2.js [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+	// Function body:
+	if ( ido === 2 ) {
+		for ( k = 1; k <= l1; ++k ) {
+			ch[ chRef( 1, k, 1, l1, ido ) - chOffset ] = cc[ ccRef( 1, 1, k, 2, ido ) - ccOffset ] + cc[ ccRef( 1, 2, k, 2, ido ) - ccOffset ];
+			ch[ chRef( 1, k, 2, l1, ido ) - chOffset ] = cc[ ccRef( 1, 1, k, 2, ido ) - ccOffset ] - cc[ ccRef( 1, 2, k, 2, ido ) - ccOffset ];
+			ch[ chRef( 2, k, 1, l1, ido ) - chOffset ] = cc[ ccRef( 2, 1, k, 2, ido ) - ccOffset ] + cc[ ccRef( 2, 2, k, 2, ido ) - ccOffset ];
+			ch[ chRef( 2, k, 2, l1, ido ) - chOffset ] = cc[ ccRef( 2, 1, k, 2, ido ) - ccOffset ] - cc[ ccRef( 2, 2, k, 2, ido ) - ccOffset ];
+		}
+	} else {
+		for ( k = 1; k <= l1; ++k ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/fft/base/fftpack/lib/passf2.js]: Same comments.

File: lib/node_modules/@stdlib/fft/base/fftpack/lib/passf3.js [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+			ch[ chRef( 2, k, 1, l1, ido ) - chOffset ] = cc[ ccRef( 2, 1, k, 3, ido ) - ccOffset ] + ti2;
+			cr3 = taui * ( cc[ ccRef( 1, 2, k, 3, ido ) - ccOffset ] - cc[ ccRef( 1, 3, k, 3, ido ) - ccOffset ] );
+			ci3 = taui * ( cc[ ccRef( 2, 2, k, 3, ido ) - ccOffset ] - cc[ ccRef( 2, 3, k, 3, ido ) - ccOffset ] );
+			ch[ chRef( 1, k, 2, l1, ido ) - chOffset ] = cr2 - ci3;
+			ch[ chRef( 1, k, 3, l1, ido ) - chOffset ] = cr2 + ci3;
+			ch[ chRef( 2, k, 2, l1, ido ) - chOffset ] = ci2 + cr3;
+			ch[ chRef( 2, k, 3, l1, ido ) - chOffset ] = ci2 - cr3;
+		}
+	} else {
+		for ( k = 1; k <= l1; ++k ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/fft/base/fftpack/lib/passf3.js]: Same comments.

File: lib/node_modules/@stdlib/fft/base/fftpack/lib/passf3.js [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+
+// MODULES //
+
+var chRef = require( './ch_ref.js' );
+var ccRef = require( './cc_ref.js' );
+
+
+// VARIABLES //
+
+var taur = -0.5;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/fft/base/fftpack/lib/passf3.js]: Do you know what the significance of these magic numbers is?

File: lib/node_modules/@stdlib/fft/base/fftpack/lib/passfb5.js [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+
+// MODULES //
+
+var chRef = require( './ch_ref.js' );
+var ccRef = require( './cc_ref.js' );
+
+
+// VARIABLES //
+
+var TR11 = 0.309016994374947;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/fft/base/fftpack/lib/passfb5.js]: Any significance to these factors?

File: lib/node_modules/@stdlib/fft/base/fftpack/lib/passfb5.js [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+	var dr4;
+	var dr5;
+	var tr2;
+	var tr3;
+	var tr4;
+	var tr5;
+	var i;
+	var k;
+
+	TI11 *= fsign;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/fft/base/fftpack/lib/passfb5.js]: Just checking. Seems a little strange to be mutating a variable in the parent scope.

File: lib/node_modules/@stdlib/fft/base/fftpack/lib/radb4.js [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+
+// MODULES //
+
+var chRef = require( './ch_ref.js' );
+var ccRef = require( './cc_ref.js' );
+
+
+// VARIABLES //
+
+var sqrt2 = 1.414213562373095;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/fft/base/fftpack/lib/radb4.js]: We should have this as a stdlib package, correct?

File: lib/node_modules/@stdlib/fft/base/fftpack/lib/radbg.js [hunk 12]
```diff
@@ -0,0 +1,300 @@
...
+* HOLDERS BE LIABLE FOR ANY CLAIM, INDIRECT, INCIDENTAL, SPECIAL,
+* EXEMPLARY, OR CONSEQUENTIAL DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+* ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+* CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH THE
+* SOFTWARE.
+* ```
+*/
+
+/* eslint-disable max-len */
+/* eslint-disable max-statements */
```
Comment (kgryte) [inline lib/node_modules/@stdlib/fft/base/fftpack/lib/radbg.js]: These can be combined into a single comment.

File: lib/node_modules/@stdlib/fft/base/fftpack/lib/radbg.js [hunk 13]
```diff
@@ -0,0 +1,300 @@
...
+
+	// Parameter adjustments...
+	chOffset = 1 + ( ido * ( 1 + l1 ) );
+	c1Offset = 1 + ( ido * ( 1 + l1 ) );
+	ccOffset = 1 + ( ido * ( 1 + ip ) );
+	ch2Offset = 1 + idl1;
+	c2Offset = 1 + idl1;
+
+	// Function body:
+	arg = ( 2 * PI ) / ( ip );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/fft/base/fftpack/lib/radbg.js]: You can use the `two-pi` constant.

File: lib/node_modules/@stdlib/fft/base/fftpack/lib/radbg.js [hunk 14]
```diff
@@ -0,0 +1,300 @@
...
+	dcp = cos( arg );
+	dsp = sin( arg );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/fft/base/fftpack/lib/radbg.js]: Instead of two calls, you can use `sincos`.

File: lib/node_modules/@stdlib/fft/base/fftpack/lib/radf4.js [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+
+// MODULES //
+
+var chRef = require( './ch_ref.js' );
+var ccRef = require( './cc_ref.js' );
+
+
+// VARIABLES //
+
+var hsqt2 = 0.7071067811865475;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/fft/base/fftpack/lib/radf4.js]: We may also have this constant: `half-sqrt-two`.

File: lib/node_modules/@stdlib/fft/base/fftpack/lib/passb2.js (hunk 3 above)
Comment (gunjjoshi) [reply lib/node_modules/@stdlib/fft/base/fftpack/lib/passb2.js]: The reference implementation uses `++k`. But, I don't think replacing it with `k++` will cause any issues. I'll change this here and everywhere.

File: lib/node_modules/@stdlib/fft/base/fftpack/lib/passf3.js (hunk 8 above)
Comment (gunjjoshi) [reply lib/node_modules/@stdlib/fft/base/fftpack/lib/passf3.js]: `taur` and `taui` are `cos( ( 2 * PI ) / 3 )` and `-sin( ( 2 * PI ) / 3 )` respectively. I'll replace these hardcoded values by these trigonometric expressions.

File: lib/node_modules/@stdlib/fft/base/fftpack/lib/passfb5.js (hunk 9 above)
Comment (gunjjoshi) [reply lib/node_modules/@stdlib/fft/base/fftpack/lib/passfb5.js]: These too, can be expressed as trigonometric expressions. For instance, `TR11` can be derived from `cos( ( 2 * PI ) / 5 )`. I'll update these accordingly.

File: lib/node_modules/@stdlib/fft/base/fftpack/lib/passfb5.js (hunk 10 above)
Comment (gunjjoshi) [reply lib/node_modules/@stdlib/fft/base/fftpack/lib/passfb5.js]: In the [reference implementation](https://github.com/marton78/pffft/blob/master/fftpack.c#L442), these constants are multiplied with `fsign` at the time of declaration itself. But, as `fsign` comes as an argument to the `passfb5` function, we can't use it in the `// VARIABLES //` section.

Comment (gunjjoshi) [reply lib/node_modules/@stdlib/fft/base/fftpack/lib/passfb5.js]: Should I move these declarations inside the function, similar to the reference implementation?

File: lib/node_modules/@stdlib/fft/base/fftpack/lib/radb4.js (hunk 11 above)
Comment (gunjjoshi) [reply lib/node_modules/@stdlib/fft/base/fftpack/lib/radb4.js]: Yes, I'll use it, thanks!

File: lib/node_modules/@stdlib/fft/base/fftpack/lib/index.js [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+var ns = {};
+
+/**
+* @name cfftb
+* @memberof ns
+* @readonly
+* @type {Function}
+* @see {@link module:@stdlib/fft/base/fftpack/cfftb}
+*/
+setReadOnly( ns, 'cfftb', require( './cfftb.js' ) );
```
Comment (gunjjoshi) [inline lib/node_modules/@stdlib/fft/base/fftpack/lib/index.js]: I've added the first function here, i.e., `cfftb`. But, is this the correct way to expose the function?

Comment (kgryte) [reply lib/node_modules/@stdlib/fft/base/fftpack/lib/index.js]: Are you wanting to export a namespace or a standalone package?

Comment (gunjjoshi) [reply lib/node_modules/@stdlib/fft/base/fftpack/lib/index.js]: We'll be exporting multiple functions from here, such as [`cfftb`](https://github.com/marton78/pffft/blob/master/fftpack.c#L1846), [`cfftf`](https://github.com/marton78/pffft/blob/master/fftpack.c#L1925), [`cffti`](https://github.com/marton78/pffft/blob/master/fftpack.c#L2027), etc. But, all of them would be in the same `fftpack` package, right? Should this be a standalone package?

Comment (kgryte) [reply lib/node_modules/@stdlib/fft/base/fftpack/lib/index.js]: Not sure. Maybe leave out the namespace setup to begin with? If we were able to split into separate packages, that would be ideal, but it really depends on how much common code all the interfaces share.

**Comment by gunjjoshi** [comment]
I had a confusion here. If we see the reference implementations of [`cosqf`](https://github.com/marton78/pffft/blob/master/fftpack.c#L2359) and [`cosqf1`](https://github.com/marton78/pffft/blob/master/fftpack.c#L2319), `cosqf` calls `cosqf1` [here](https://github.com/marton78/pffft/blob/master/fftpack.c#L2375), and passes pointers to the `wsave` arrays, as `cosqf1` accepts pointers to that array. 
But in the current JS implementation in this PR, I am passing a single element of the array [here](https://github.com/stdlib-js/stdlib/pull/4121/files#diff-639ebae88faa3cee116421e75c70226ad709625f063d3fda1a5ff9fefa09d09eR86), while `cosqf1` takes an entire array as an argument [here](https://github.com/stdlib-js/stdlib/pull/4121/files#diff-416134bb617cfe2d90931335cdf67ed55e76d5c158391d46273fc4db4bad1b1dR78). This might not be the right approach. Is there any other way to achieve the same behavior as shown in the reference C implementation?

**Comment by gunjjoshi** [comment]
Should I follow something like what we are doing in the [C](https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/math/base/special/modf/src/main.c#L45) and [JS](https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/math/base/special/modf/lib/assign.js#L57) implementations of [modf](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/math/base/special/modf)?

**Comment by kgryte** [comment]
@gunjjoshi It's a little hard to follow what you are asking given the various links. Can you provide relevant snippets in a single comment so we can see what you are referencing in one view?

**Comment by gunjjoshi** [comment]
@kgryte Ahh, sorry for the confusion.

This is the `cosqf1` function:

```c
static void cosqf1(integer n, real *x, real *w, real *xh)
```

This `cosqf1` is used within the `cosqf` function as:

```c
cosqf1(n, &x[1], &wsave[1], &wsave[n + 1]);
```

In order to replicate this same behavior in JS, currently, what I'm doing is:

`cosqf1`:

```js
function cosqf1( n, x, w, xh )
```
where `n` is a `number`, while `x`, `w` and `xh` are `Float64Array`s.

And, while calling this `cosqf1` from `cosqf`, we are currently using it as:

```js
cosqf1( n, x[ 1 - 1 ], wsave[ 1 - 1 ], wsave[ n + 1 - 1 ] );
```

But in this JS version, instead of passing arrays, I am currently passing single elements, such as `wsave[ n + 1 - 1 ]`.
This seems to be wrong. Is there any other way of passing arrays here, but from a specific index (something like a subarray)?

I think this problem might be summed up as, how should we replicate the logic of passing subarrays (using pointers) in C, for the JS implementation?

**Comment by kgryte** [comment]
@gunjjoshi What about providing an array and then offsets, similar in concept to the ndarray APIs we have. E.g.,

```javascript
xptr = ...;
wptr = ...;
xhptr = ...;

cosqf1(n, x, xptr, wsave, wptr, xhptr ); // =>cosqf1( n: number, x: Float64Array, xptr: number, wsave: Float64Array, wptr: number, xhptr: number )
```

where the `*ptr` variables are indices into the respective arrays.

Another alternative is that you pass down slices which has more overhead due to object creation.

```
cosqf1( n, x.slice( 1 ), wsave.slice( 1 ), wsave.slice( n+1 ) )
```

**Comment by gunjjoshi** [comment]
Thanks, @kgryte, I'll go with the first method!

**Comment by gunjjoshi** [comment]
The required functions are implemented now, will move on to testing the implementations locally, comparing with the reference ones and then creating the tests.

**Comment by kgryte** [comment]
> The required functions are implemented now, will move on to testing the implementations locally, comparing with the reference ones and then creating the tests.

Great! Thanks for your continued work on this!

**Comment by gunjjoshi** [comment]
@kgryte For the tests, should we have the tests for all functions in a single `test.js` file, or should we keep them separate, such as `test.rffti.js`, `test.cosqb.js`, etc.? The functions that would be tested are mentioned here: https://github.com/marton78/pffft/blob/master/fftpack.c#L2710

**Comment by kgryte** [comment]
@gunjjoshi Let's try in separate files to start.

**Comment by gunjjoshi** [comment]
The original C reference implementation here: https://github.com/marton78/pffft/blob/master/fftpack.c has a tolerance of `0.001`. But currently, I'm getting it as follows:

```
      ✖ forward transform error: 0.4907203841224826 (expected <= 0.001)
      ✖ backward transform error: 40.50771641905412 (expected <= 0.001)
      ✖ forward-backward transform error: 34.05132902258452 (expected <= 0.001)
```

Seems like a major difference. I went through the whole implementation multiple times, still trying to figure out the problem. Especially, testing [`rffti`](https://github.com/stdlib-js/stdlib/pull/4121/files#diff-b677dfe331e6afea0d2fe05b6f2fd1d5d5196e21e23308898bf8194902c82297), [`rfftf`](https://github.com/stdlib-js/stdlib/pull/4121/files#diff-07a2b0a08b7395cd027a963477df00d66c20b74eeb1c52182279316eaf7be4af) and [`rfftb`](https://github.com/stdlib-js/stdlib/pull/4121/files#diff-389bb50b04ae5ec89a5fded3a38f943b16a15bb8697f22fb6098f2096a9c43cc) implementations.

File: lib/node_modules/@stdlib/fft/base/fftpack/test/test.rfft.js [hunk 17]
```diff
@@ -0,0 +1,185 @@
...
+var pow = require( '@stdlib/math/base/special/pow' );
+var floor = require( '@stdlib/math/base/special/floor' );
+var rffti = require( './../lib/rffti.js' );
+var rfftf = require( './../lib/rfftf.js' );
+var rfftb = require( './../lib/rfftb.js' );
+
+
+// VARIABLES //
+
+var nd = [ 120, 91, 54, 49, 32, 28, 24, 8, 4, 3, 2 ];
```
Comment (gunjjoshi) [inline lib/node_modules/@stdlib/fft/base/fftpack/test/test.rfft.js:40]: The above error values are for `120`, as I am testing each element one by one, for simplicity. While testing with smaller values such as `2` and `3`, these tolerances are much closer to the the expected ones, but still not acceptable.

**Comment by gunjjoshi** [comment]
Here's what I've done so far:

- Implemented all the functions, as given in the reference C implementation: https://github.com/marton78/pffft/blob/master/fftpack.c
- Added tests for [`rffti`](https://github.com/stdlib-js/stdlib/pull/4121/files#diff-b677dfe331e6afea0d2fe05b6f2fd1d5d5196e21e23308898bf8194902c82297), [`rfftf`](https://github.com/stdlib-js/stdlib/pull/4121/files#diff-07a2b0a08b7395cd027a963477df00d66c20b74eeb1c52182279316eaf7be4af) and [`rfftb`](https://github.com/stdlib-js/stdlib/pull/4121/files#diff-389bb50b04ae5ec89a5fded3a38f943b16a15bb8697f22fb6098f2096a9c43cc), in [`test.rfft.js`](https://github.com/stdlib-js/stdlib/pull/4121/files#diff-a7c3eb218d26381a783ad77dd8a8dcffb2ae68b8c55a3b4d4ece66b7731f3371). I've implemented the same tests given in the reference implementation here: https://github.com/marton78/pffft/blob/master/fftpack.c#L2656
- Used offsets in JS to replicate the behavior of pointers in C.
- Some tests still fail, in [`test.rfft.js`](https://github.com/stdlib-js/stdlib/pull/4121/files#diff-a7c3eb218d26381a783ad77dd8a8dcffb2ae68b8c55a3b4d4ece66b7731f3371).

File: lib/node_modules/@stdlib/fft/base/fftpack/lib/rffti.js [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+*/
+function rffti( n, wsave, wptr ) {
+	// Parameter adjustments:
+	wptr -= 1;
+
+	// Function Body:
+	if ( n === 1 ) {
+		return;
+	}
+	rffti1( n, wsave, n + 1 + wptr, wsave, 1 + ( n << 1 ) + wptr );
```
Comment (gunjjoshi) [inline lib/node_modules/@stdlib/fft/base/fftpack/lib/rffti.js]: The C reference for this part calls the `rffti1` function as: https://github.com/marton78/pffft/blob/master/fftpack.c#L2251

Just wanted to confirm, whether the use of offsets here is the correct C port of the pointers given in the reference implementation?

File: lib/node_modules/@stdlib/fft/base/fftpack/rffti/docs/repl.txt [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+
+    Examples
+    --------
+    > var N = 8;
+    > var workspace = new {{alias:@stdlib/array/float64}}( (2*N) + 34 );
+    > {{alias}}( N, workspace, 1, 0 )
+    > workspace.slice()
+    <Float64Array>[ 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0.7071067811865476,
+    ... 0.7071067811865475, 0, 0, 0, 0, 8, 2, 2, 4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    ... 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 ]
```
Comment (gunjjoshi) [inline lib/node_modules/@stdlib/fft/base/fftpack/rffti/docs/repl.txt]: This was giving linting error. How can I display this array, which is more than 80 characters long and contains 50 elements?

File: lib/node_modules/@stdlib/fft/base/fftpack/rffti/docs/repl.txt [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+    Returns
+    -------
+    void
+
+    Examples
+    --------
+    > var N = 8;
+    > var workspace = new {{alias:@stdlib/array/float64}}( (2*N) + 34 );
+    > {{alias}}( N, workspace, 1, 0 )
+    > workspace.slice()
```
Comment (kgryte) [inline lib/node_modules/@stdlib/fft/base/fftpack/rffti/docs/repl.txt]:
```suggestion
    > workspace
```

File: lib/node_modules/@stdlib/fft/base/fftpack/rffti/docs/repl.txt (hunk 19 above)
Comment (kgryte) [inline lib/node_modules/@stdlib/fft/base/fftpack/rffti/docs/repl.txt]:
```suggestion
    e.g., <Float64Array>[ ..., ~0.707, ... ~0.707, ..., 8, 2, 2, 4, ... ]
```

**Comment by kgryte** [comment]
@gunjjoshi As we merge packages in upstream, I suggest keeping this PR up-to-date. E.g., removing the `decompose` package in this PR, merging in the latest, and then updating all usages of `decompose` to match the API in the latest `develop`. Over time, this should mean that the number of files changed in this PR decreases and can help us keep tabs on how much work still remains.
